satellite/satellitedb/offers.go

```diff
@@ -0,0 +1,163 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information
+
+package satellitedb
+
+import (
+	"context"
+	"database/sql"
+	"time"
+
+	"github.com/zeebo/errs"
+
+	"storj.io/storj/satellite/marketing"
+	dbx "storj.io/storj/satellite/satellitedb/dbx"
+)
+
+type offers struct {
+	db *dbx.DB
+}
+
+// ListAll returns all offers from the db
+func (offers *offers) ListAll(ctx context.Context) ([]marketing.Offer, error) {
+	offersDbx, err := offers.db.All_Offer(ctx)
+	if err != nil {
+		return nil, marketing.OffersErr.Wrap(err)
+	}
+
+	return offersFromDBX(offersDbx)
+}
+
+// GetCurrent returns an offer that has not expired based on offer status
+func (offers *offers) GetCurrent(ctx context.Context, offerStatus marketing.OfferStatus, offerType marketing.OfferType) (*marketing.Offer, error) {
+	var statement string
+	const columns = "id, name, description, award_credit_in_cents, invitee_credit_in_cents, award_credit_duration_days, invitee_credit_duration_days, redeemable_cap, num_redeemed, expires_at, created_at, status, type"
+	statement = `
+		WITH o AS (
+			SELECT ` + columns + ` FROM offers WHERE status=? AND type=? AND expires_at>? AND num_redeemed < redeemable_cap
+		)
+		SELECT ` + columns + ` FROM o
+		UNION ALL
+		SELECT ` + columns + ` FROM offers
+		WHERE type=? AND status=?
+		AND NOT EXISTS (
+			SELECT id FROM o
+		) order by created_at desc;`
+
+	rows := offers.db.DB.QueryRowContext(ctx, offers.db.Rebind(statement), offerStatus, offerType, time.Now().UTC(), offerType, marketing.Default)
+
+	o := marketing.Offer{}
+	err := rows.Scan(&o.ID, &o.Name, &o.Description, &o.AwardCreditInCents, &o.InviteeCreditInCents, &o.AwardCreditDurationDays, &o.InviteeCreditDurationDays, &o.RedeemableCap, &o.NumRedeemed, &o.ExpiresAt, &o.CreatedAt, &o.Status, &o.Type)
+	if err == sql.ErrNoRows {
+		return nil, marketing.OffersErr.New("no current offer")
+	}
+	if err != nil {
+		return nil, marketing.OffersErr.Wrap(err)
+	}
+
+	return &o, nil
+}
+
+// Create inserts a new offer into the db
+func (offers *offers) Create(ctx context.Context, o *marketing.NewOffer) (*marketing.Offer, error) {
+	currentTime := time.Now()
+	if o.ExpiresAt.Before(currentTime) {
+		return nil, marketing.OffersErr.New("expiration time: %v can't be before: %v", o.ExpiresAt, currentTime)
+	}
+
+	tx, err := offers.db.Open(ctx)
+	if err != nil {
+		return nil, marketing.OffersErr.Wrap(err)
+	}
+
+	statement := offers.db.Rebind(`
+		UPDATE offers SET status=?, expires_at=?
+		WHERE status=? AND type=? AND expires_at>?;
+	`)
+	_, err = tx.Tx.ExecContext(ctx, statement, marketing.Done, currentTime, o.Status, o.Type, currentTime)
```

Comment:
I'm not sure I understand what's happening with this sql statement. I know its updating but can you add a comment why?

Reply:
I added a comment for the statment

Reply:
Typo in the comment